### PR TITLE
adding stabs for unsupported whitelisted functions

### PR DESCRIPTION
**Pull request by astrophysik**
General
-----------

The main purpose of this pr is to change the behavior of kphp light when using unsupported standard functions.
From:
```
$res = not_supported_builtin_yet(); // compile error
``` 
To:
```
$res = not_supported_builtin_yet(); // runtime error
``` 

Details
---------

Builtins divided into two groups - template and non template. For non template functions added tag `@generate-stub`. It ask compiler to generate body of function with call to `php_critical_error`. For template functions I add stubs in runtime-light headers.

Note that this pr doesn't add support for all builtins.

### CODE REVIEW

File: runtime-light/stdlib/array-functions.h [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+#pragma once
+
+#include "runtime-core/runtime-core.h"
+
+constexpr int64_t SORT_REGULAR = 0;
+constexpr int64_t SORT_NUMERIC = 1;
+constexpr int64_t SORT_STRING = 2;
+
+template<class T>
+inline string f$implode(const string &s, const array<T> &a) {
```
Comment (apolyakov) [inline runtime-light/stdlib/array-functions.h:14]: Template function definitions are implicitly `inline`. So it's useless to explicitly mark them inline.

File: runtime-light/stdlib/regexp.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2024 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+
+#pragma once
+
+#include "runtime-core/runtime-core.h"
+
+inline  int64_t preg_replace_count_dummy;
```
Comment (apolyakov) [inline runtime-light/stdlib/regexp.h:10]: Global variable?

**Review by apolyakov — CHANGES_REQUESTED** [review]
Let's not just add random header files directly to `stdlib` folder. Instead, let's create directory structure, e.g. `stdlib/array`, `stdlib/string` etc

File: compiler/code-gen/vertex-compiler.cpp [hunk 3]
```diff
@@ -1364,6 +1364,14 @@ bool compile_tracing_profiler(FunctionPtr func, CodeGenerator &W) {
   return true;
 }
 
+void compile_generated_stub(VertexAdaptor<op_function> func_root, CodeGenerator &W) {
+  FunctionPtr func = func_root->func_id;
+  W << FunctionDeclaration(func, false) << " " <<
+    BEGIN;
+  W << "php_critical_error(\"call to unsupported function\");" << NL;
```
Comment (mkornaukhov03) [inline compiler/code-gen/vertex-compiler.cpp:1371]: I think it's better to write function name here

File: compiler/pipes/collect-main-edges.cpp [hunk 4]
```diff
@@ -328,7 +328,7 @@ void CollectMainEdgesPass::on_func_call(VertexAdaptor<op_func_call> call) {
     VertexPtr arg = call->args()[i];
     auto param = function_params[i].as<op_func_param>();
     bool is_cast_param = !stage::get_function()->file_id->is_strict_types && param->is_cast_param;
-
+    //
```
Comment (mkornaukhov03) [inline compiler/pipes/collect-main-edges.cpp:331]: Redundant comment

File: compiler/pipes/final-check.cpp [hunk 5]
```diff
@@ -358,7 +358,10 @@ void check_func_call_params(VertexAdaptor<op_func_call> call) {
     auto param = func_params[i].as<op_func_param>();
     bool is_callback_passed_to_extern = f->is_extern() && param->type_hint && param->type_hint->try_as<TypeHintCallable>();
     if (!is_callback_passed_to_extern) {
-      kphp_error(call_params[i]->type() != op_callback_of_builtin, "Unexpected function pointer");
+      if (call_params[i]->type() == op_callback_of_builtin) {
+        kphp_notice(fmt_format("function {} is extern {}", f->name, f->is_extern()));
+        kphp_error(call_params[i]->type() != op_callback_of_builtin, "Unexpected function pointer");
```
Comment (mkornaukhov03) [inline compiler/pipes/final-check.cpp:363]: Looks like debug print

File: runtime-light/stdlib/instance-cache/instance-cache.h [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+template<typename ClassInstanceType>
+bool f$instance_cache_store(const string &key, const ClassInstanceType &instance, int64_t ttl = 0) {
+  php_critical_error("call to unsupported function");
+}
+
+template<typename ClassInstanceType>
+ClassInstanceType f$instance_cache_fetch(const string &class_name, const string &key, bool even_if_expired = false) {
+  php_critical_error("call to unsupported function");
+
+}
```
Comment (mkornaukhov03) [inline runtime-light/stdlib/instance-cache/instance-cache.h:18]: Let's add new line at end of each file

**Review by mkornaukhov03 — CHANGES_REQUESTED** [review]

File: runtime-light/stdlib/regexp/regexp.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2024 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+
+#pragma once
+
+#include "runtime-core/runtime-core.h"
+
+struct RegexpContext {
```
Comment (apolyakov) [inline runtime-light/stdlib/regexp/regexp.h:10]: Let's keep using our naming: `RegexComponentState`

File: runtime-light/stdlib/regexp/regexp.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  int64_t preg_replace_count_dummy;
```
Comment (apolyakov) [inline runtime-light/stdlib/regexp/regexp.h:11]: `int64_t preg_replace_count_dummy{}`

File: runtime-light/stdlib/regexp/regexp.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+  static RegexpContext& get() noexcept;
+};
+
+
+template<class T1, class T2, class T3, class = enable_if_t_is_optional<T3>>
+inline auto f$preg_replace(const T1 &, const T2 &, const T3 &, int64_t  = -1, int64_t & = RegexpContext::get().preg_replace_count_dummy) {
```
Comment (apolyakov) [inline runtime-light/stdlib/regexp/regexp.h:18]: Same problem with `inline`. Please, also check other files.

File: runtime-light/stdlib/stdlib.cmake [hunk 10]
```diff
@@ -15,4 +15,5 @@ prepend(
   rpc/rpc-extra-info.cpp
   rpc/rpc-tl-error.cpp
   rpc/rpc-tl-query.cpp
-  rpc/rpc-tl-request.cpp)
+  rpc/rpc-tl-request.cpp
+        regexp/regexp.cpp)
```
Comment (apolyakov) [inline runtime-light/stdlib/stdlib.cmake:19]: Formatting issue

File: runtime-light/stdlib/stdlib.cmake [hunk 11]
```diff
@@ -1,16 +1,17 @@
...
+        output/print-functions.cpp
+        rpc/rpc-api.cpp
+        rpc/rpc-context.cpp
+        rpc/rpc-extra-headers.cpp
+        rpc/rpc-extra-info.cpp
+        rpc/rpc-tl-error.cpp
+        rpc/rpc-tl-query.cpp
+        rpc/rpc-tl-request.cpp
+        string/concat.cpp
+        regex/regex-functions.cpp)
```
Comment (mkornaukhov03) [inline runtime-light/stdlib/stdlib.cmake:17]: No extra line in the end. Please, add it but up to you

**Review by mkornaukhov03 — CHANGES_REQUESTED** [review]
Does this PR provides unsupported classes?
It have Compiler Error now:
```php
<?php

$a = new Exception("kek", 42);
```
G++ error:
```
path/to/kphp/dev_kphp/kphp_out/kphp/o_globals_reset/globals_reset.0.cpp:8:52: error: use of undeclared identifier 'C$Exception'
    8 |   hard_reset_var((*reinterpret_cast<class_instance<C$Exception>*>(php_globals.mem()+112)));
```

Would it be fixed in new PR or in this one?

**Comment by astrophysik** [comment]
> Does this PR provides unsupported classes? I have Compiler Error now:
> 
> ```
> <?php
> 
> $a = new Exception("kek", 42);
> ```
> 
> G++ error:
> 
> ```
> path/to/kphp/dev_kphp/kphp_out/kphp/o_globals_reset/globals_reset.0.cpp:8:52: error: use of undeclared identifier 'C$Exception'
>     8 |   hard_reset_var((*reinterpret_cast<class_instance<C$Exception>*>(php_globals.mem()+112)));
> ```
> 
> Would it be fixed in new PR or in this one?

Support for internal Classes and Visitors will be in next pr

**Review by mkornaukhov03 — APPROVED** [review]
lgtm

**Review by apolyakov — APPROVED** [review]

**Pull request merged**